pkg/controller/route53hostedzone/zonereconciler.go

```diff
@@ -0,0 +1,336 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package route53hostedzone
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/aws/aws-sdk-go/service/route53"
+	cov1 "github.com/openshift/cluster-operator/pkg/apis/clusteroperator/v1alpha1"
+	lister "github.com/openshift/cluster-operator/pkg/client/listers_generated/clusteroperator/v1alpha1"
+	clustopaws "github.com/openshift/cluster-operator/pkg/clusterapi/aws"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeclientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/aws/aws-sdk-go/aws"
+	controller "github.com/openshift/cluster-operator/pkg/controller"
+
+	coclient "github.com/openshift/cluster-operator/pkg/client/clientset_generated/clientset"
+)
+
+// ZoneReconciler manages getting the desired state, getting the current state and reconciling the two.
+type ZoneReconciler struct {
+	desiredState   *cov1.DNSZone
+	currentState   *route53.HostedZone
+	dnsZonesLister lister.DNSZoneLister
+
+	// logrus logger
+	logger log.FieldLogger
+
+	// Spec stuff
+	dnszone string
+
+	// Kube related fields
+	name      string
+	namespace string
+
+	// AWS related fields
+	coAWSClient clustopaws.Client
+
+	// Kubernetes client
+	kubeClient kubeclientset.Interface
+
+	// Cluster Operator Client for where cluster operator objects are stored.
+	client coclient.Interface
+
+	// Function pointers, useful for dependency injection
+	funcs *functionPointers
+}
+
+type functionPointers struct {
+	// Function pointers, useful for dependency injection
+	getDesiredState         func() error
+	getCOAWSClient          func() error
+	getCurrentState         func() error
+	createRoute53HostedZone func() error
+	deleteRoute53HostedZone func() error
+
+	// Function that builds the aws client
+	awsClientBuilder func(kubeClient kubeclientset.Interface, secretName, namespace, region string) (clustopaws.Client, error)
+}
+
+// NewZoneReconcilerInput manages getting the desired state, getting the current state and reconciling the two.
+type NewZoneReconcilerInput struct {
+	// The DNSZone key that is being synced
+	key string
+
+	// The kubernetes zone lister that was created for zone objects.
+	dnsZonesLister lister.DNSZoneLister
+
+	// The kubernetes client to use
+	kubeClient kubeclientset.Interface
+
+	// Cluster Operator Client for where cluster operator objects are stored.
+	client coclient.Interface
+
+	// Function pointers, useful for dependency injection
+	functionOverrides *functionPointers
+
+	// logrus logger
+	logger log.FieldLogger
+}
+
+// NewZoneReconciler creates a new ZoneReconciler object. A new ZoneReconciler is expected to be created for each controller sync.
+func NewZoneReconciler(newZoneReconcilerInput *NewZoneReconcilerInput) (*ZoneReconciler, error) {
```

Comment:
I'm not sure what you're saying here.

Are you saying the name itself is AWS'y?  If so, I'm open to changing it. How about `ZoneReconcilerOptions` ?

Or, are you saying that having a struct that contains the function params doesn't fit in with CO / Kube?

If that's the case, I believe it does fit in (at least from the small exposure I've had to kube code).

Here are some examples:
[ListOptions](https://github.com/openshift/cluster-operator/blob/master/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L318)
[DeleteOptions](https://github.com/openshift/cluster-operator/blob/master/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L423)
[ExportOptions](https://github.com/openshift/cluster-operator/blob/master/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L379)

I may be misunderstanding this kube pattern, of course. If I am, please let me know.

I'm open to changing this to make it better / fit in, I just don't know what you're looking for.

Reply:
Simplicity and consistency is what I'm looking for. Effectively I'm looking for the reason why we would use an object here instead of just passing params directly to the function? At present this isn't really something we've done in our codebase as far as I can recall. 

Those Kube examples are interesting, I don't know for sure the thought process behind why they did that, but I suspect the pattern may be intended for functions which (1) are called hundreds or thousands of times over, (2) all arguments are optional and many callers don't need to pass any at all, (3) they'd like to add more arguments in the future without having to update thousands of method calls.

Those don't really seem to apply here, your args all look mandatory to me and if more were added in the future, you would lose compile time notification that code needs to be updated, you'd have to wait until something broke or hope your unit tests covered them all.